### PR TITLE
Specify minimum price of product and apply it on catalogPromotions

### PR DESCRIPTION
**Pull request by SirDomin**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

CatalogPromotionApplicator uses minimum price

### CODE REVIEW

File: features/product/managing_product_variants/adding_product_variant.feature [hunk 1]
```diff
@@ -89,3 +89,15 @@ Feature: Adding a new product variant
         Then I should be notified that it has been successfully created
         And the "VODKA_WYBOROWA_PREMIUM" variant of the "Wyborowa Vodka" product should appear in the store
         And the variant with code "VODKA_WYBOROWA_PREMIUM" should be priced at $0.00 for channel "United States"
+
+    @ui
```
Comment (AdamKasp) [inline features/product/managing_product_variants/adding_product_variant.feature:93]:
```suggestion
    @api @ui
```

**Review by AdamKasp — APPROVED** [review]

File: features/product/managing_product_variants/adding_product_variant.feature (hunk 1 above)
Comment (SirDomin) [reply features/product/managing_product_variants/adding_product_variant.feature:93]: yeah, this part is still wip as there's no api steps implemented for this feature file

File: features/promotion/applying_catalog_promotions/applying_catalog_promotions_with_minimum_price.feature [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+Feature: Applying catalog promotions with minimum price
+    In order to avoid too many promotions applied on product
+    As a Visitor
+    I want to see discounted products in the catalog up to minimum price
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a "T-Shirt" configurable product
+        And this product has "PHP T-Shirt" variant priced at "$20.00"
+        And "PHP T-Shirt" variant has minimum price "$15.00" in "United States" channel
```
Comment (GSadee) [inline features/promotion/applying_catalog_promotions/applying_catalog_promotions_with_minimum_price.feature:11]:
```suggestion
        And the "PHP T-Shirt" variant has minimum price "$15.00" in "United States" channel
```

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 3]
```diff
@@ -128,6 +128,14 @@ public function iSetItsPriceTo(?string $price = null, ?ChannelInterface $channel
         $this->createPage->specifyPrice($price ?? '', $channel ?? $this->sharedStorage->get('channel'));
     }
 
+    /**
+     * @When /^I set its minimum price to "(?:€|£|\$)([^"]+)" for ("([^"]+)" channel)$/
+     */
+    public function iSetItsMinimumPriceTo(string $price, ChannelInterface $channel)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php:134]:
```suggestion
    public function iSetItsMinimumPriceTo(string $price, ChannelInterface $channel): void
```

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 4]
```diff
@@ -226,6 +234,16 @@ public function theVariantWithCodeShouldBePricedAtForChannel(ProductVariantInter
         Assert::same($this->updatePage->getPriceForChannel($channel), $price);
     }
 
+    /**
+     * @Then /^the (variant with code "[^"]+") should have minimum price (?:€|£|\$)([^"]+) for (channel "([^"]+)")$/
+     */
+    public function theVariantWithCodeShouldHaveMinimumPriceForChannel(ProductVariantInterface $productVariant, string $price, ChannelInterface $channel)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php:240]:
```suggestion
    public function theVariantWithCodeShouldHaveMinimumPriceForChannel(ProductVariantInterface $productVariant, string $price, ChannelInterface $channel): void
```

File: src/Sylius/Behat/Page/Admin/ProductVariant/CreatePage.php [hunk 5]
```diff
@@ -113,6 +118,7 @@ protected function getDefinedElements(): array
             'shipping_category' => '#sylius_product_variant_shippingCategory',
             'shipping_required' => '#sylius_product_variant_shippingRequired',
             'original_price' => '#sylius_product_variant_channelPricings input[name$="[originalPrice]"][id*="%channelCode%"]',
+            'minimum_price' => '#sylius_product_variant_channelPricings input[name$="[minimumPrice]"][id*="%channelCode%"]',
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Admin/ProductVariant/CreatePage.php:121]: Wrong order of keys

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php [hunk 6]
```diff
@@ -13,18 +13,80 @@
...
+     */
+    public function iSpecifyItsCodeAs(string $code): void
+    {
+        $this->client->addRequestData('code', $code);
+    }
+
+    /**
+     * @When /^I set its price to ("[^"]+") for ("[^"]+" channel)$/
+     */
+    public function iSetItsPriceToForChannel(int $price, ChannelInterface $channel)
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php:63]:
```suggestion
    public function iSetItsPriceToForChannel(int $price, ChannelInterface $channel): void
```

here and in functions below

**Review by lchrusciel — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php [hunk 7]
```diff
@@ -13,18 +13,80 @@
...
+        ResponseCheckerInterface $responseChecker,
+        IriConverterInterface $iriConverter
+    ) {
         $this->client = $client;
+        $this->responseChecker = $responseChecker;
+        $this->iriConverter = $iriConverter;
+    }
+
+    /**
+     * @Given /^I want to create a new variant of (this product)$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php:44]:
```suggestion
     * @When /^I want to create a new variant of (this product)$/
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php [hunk 8]
```diff
@@ -59,6 +121,47 @@ public function iRemoveTheOriginalPriceOfTheProductVariantInChannel(
...
+    }
+
+    /**
+     * @Then /^the (variant with code "[^"]+") should be priced at ("[^"]+") for (channel "([^"]+)")$/
+     */
+    public function theVariantWithCodeShouldBePricedAtForChannel(ProductVariantInterface $productVariant, int $price, ChannelInterface $channel): void
+    {
+        $response = $this->responseChecker->getCollection($this->client->index());
+
+        Assert::true($response[0]['channelPricings'][$channel->getCode()]['price'] === $price);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php:152]: Use `Assert::same` here

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php [hunk 9]
```diff
@@ -59,6 +121,47 @@ public function iRemoveTheOriginalPriceOfTheProductVariantInChannel(
...
+    }
+
+    /**
+     * @Then /^the (variant with code "[^"]+") should have minimum price ("[^"]+") for (channel "([^"]+)")$/
+     */
+    public function theVariantWithCodeShouldHaveMinimumPriceForChannel(ProductVariantInterface $productVariant, int $minimumPrice, ChannelInterface $channel): void
+    {
+        $response = $this->responseChecker->getCollection($this->client->index());
+
+        Assert::true($response[0]['channelPricings'][$channel->getCode()]['minimumPrice'] === $minimumPrice);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php:162]: As above

File: src/Sylius/Behat/Page/Admin/ProductVariant/UpdatePage.php [hunk 10]
```diff
@@ -130,6 +135,7 @@ protected function getDefinedElements(): array
             'on_hand' => '#sylius_product_variant_onHand',
             'option_values' => '#sylius_product_variant_optionValues_%optionName%',
             'original_price' => '#sylius_product_variant_channelPricings input[name$="[originalPrice]"][id*="%channelCode%"]',
+            'minimum_price' => '#sylius_product_variant_channelPricings input[name$="[minimumPrice]"][id*="%channelCode%"]',
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Admin/ProductVariant/UpdatePage.php:138]: Wrong order of keys

File: src/Sylius/Behat/Resources/config/suites/api/product/managing_product_variants.yml [hunk 11]
```diff
@@ -7,15 +7,34 @@ default:
...
+                - sylius.behat.context.setup.geographical
                 - sylius.behat.context.setup.channel
+                - sylius.behat.context.setup.currency
+                - sylius.behat.context.setup.locale
                 - sylius.behat.context.setup.product
-                - sylius.behat.context.setup.admin_security
+                - sylius.behat.context.setup.admin_api_security
+                - sylius.behat.context.setup.taxonomy
 
                 - sylius.behat.context.api.admin.ajax
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/suites/api/product/managing_product_variants.yml:36]: Should be removed

File: src/Sylius/Behat/Resources/config/suites/api/product/managing_product_variants_ajax.yml [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+                - sylius.behat.context.setup.geographical
+                - sylius.behat.context.setup.channel
+                - sylius.behat.context.setup.currency
+                - sylius.behat.context.setup.locale
+                - sylius.behat.context.setup.product
+                - sylius.behat.context.setup.admin_security
+                - sylius.behat.context.setup.taxonomy
+
+                - sylius.behat.context.api.admin.ajax
+                - sylius.behat.context.api.admin.managing_product_variants
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/suites/api/product/managing_product_variants_ajax.yml:37]: Leave only needed contexts here

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/ProductVariant.xml [hunk 13]
```diff
@@ -22,14 +22,17 @@
...
             <group>shop:product_variant:read</group>
         </attribute>
         <attribute name="product">
             <group>admin:product_variant:read</group>
+            <group>admin:product_variant:create</group>
             <group>shop:product_variant:read</group>
         </attribute>
         <attribute name="translations">
             <group>shop:order:account:read</group>
+            <group>admin:product_variant:create</group>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/ProductVariant.xml:35]: Is it needed in this PR?

**Review by GSadee — COMMENTED** [review]
You should add also a contract test for simple product variant creation in this PR

File: src/Sylius/Behat/Resources/config/suites/api/product/managing_product_variants.yml [hunk 14]
```diff
@@ -7,15 +7,33 @@ default:
             contexts:
                 - sylius.behat.context.hook.doctrine_orm
 
+                - sylius.behat.context.transform.channel
```
Comment (AdamKasp) [inline src/Sylius/Behat/Resources/config/suites/api/product/managing_product_variants.yml:10]: do you need all of these contexts?

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, @SirDomin! :1st_place_medal:

File: features/product/managing_product_variants/autocomplete_product_variant_descriptor.feature [hunk 15]
```diff
@@ -1,4 +1,4 @@
-@managing_product_variants
+@managing_product_variants_ajax
```
Comment (Zales0123) [inline features/product/managing_product_variants/autocomplete_product_variant_descriptor.feature:1]: Why 😢

**Pull request merged**